osutil/user/user.go

```diff
@@ -0,0 +1,281 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2017 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// osutil.user implements a local user lookup and iteration module.
+//
+// Right now it's very dumb, and will not be happy on systems with a
+// lot of users. These things will get fixed, but the API this exposes
+// should survive that fix.
+package user
+
+import (
+	"bufio"
+	"bytes"
+	"errors"
+	"os"
+	"path/filepath"
+	"strconv"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/osutil/sys"
+	"github.com/snapcore/snapd/release"
+	"github.com/snapcore/snapd/strutil"
+)
+
+var minUserUID sys.UserID
+var shells = []string{"/bin/sh"}
+
+func init() {
+	Init()
+}
+
+func Init() {
```

Comment:
this looks weird and has no doc comment?  why not export_test it?

Reply:
So... I made `Init` public because if a test needs to have a user exist, it needs to call `Init` after creating /etc/passwd in the fake root; normal use doesn't need it.

Is that reasonable (in which case I do need to add a comment), or should I leave it private and let future people figure this out?

Reply:
it's not unreasonable but then two points:

* yes, it needs a comment
* a name that makes it clear it can be called again (within the constrain of not being concurrency safe) would be better maybe

Reply:
you know me and names... suggestions?

Reply:
the public use is more like a reset/re-read/re-init than Init would imply. Maybe `Reset()`?

Reply:
Reset? Reinit?

Reply:
Maybe I should make it return a `func()`, and call it `MockSomething`?

Reply:
we do have test only functions (at least when called from outside) that are not Mock, just a matter of having a good comment